### PR TITLE
Red: Chapter 5 exercises

### PR DESCRIPTION
**Pull request by utopia-2**

### CODE REVIEW

File: src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+	@Override
+	public Function<WorldMap, WorldMap> finisher() {
+		return Function.identity();
+	}
+
+	@Override
+	public Set<Characteristics> characteristics() {
+		return Stream.of( Characteristics.IDENTITY_FINISH,
+			Characteristics.UNORDERED ).collect( Collectors.toSet() );
```
Comment (a-panayotopoulos) [inline src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java:43]: I'm unsure of whether it's also `Characteristics.CONCURRENT` or not.

File: src/red/java/com/lambdaherding/edi/red/ch05/Fibonacci.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+	private final Map<Integer, Long> cache = new HashMap<>();
+
+	public Fibonacci() {
+		cache.put( 0, 0L );
+		cache.put( 1, 1L );
+	}
+
+	public long fib( int nth ) {
+		return cache.computeIfAbsent( nth, n -> fib(n - 1) + fib(n - 2) );
```
Comment (a-panayotopoulos) [inline src/red/java/com/lambdaherding/edi/red/ch05/Fibonacci.java:16]: I could stack-overflow your machine with `fib( -1 )` :smile_cat:

Comment (utopia-2) [reply src/red/java/com/lambdaherding/edi/red/ch05/Fibonacci.java:16]: Eh, it didn't ask for error checking :wink:.

File: src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java (hunk 1 above)
Comment (utopia-2) [reply src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java:43]: `WorldMap` puts to an internal `HashMap`, which can't have concurrent put ops. As it's the accumulator that'd be run concurrently, I don't think it's safe:

> Indicates that this collector is concurrent, meaning that the result container can support the accumulator function being called concurrently with the same result container from multiple threads

Comment (a-panayotopoulos) [reply src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java:43]: Cool, so if I'd used `ConcurrentHashMap` it would have been applicable?

Comment (utopia-2) [reply src/red/java/com/lambdaherding/edi/red/ch05/WorldMapCollector.java:43]: Yeah, that's right.

**Pull request merged**